src/units/abilities.cpp

```diff
@@ -600,28 +604,29 @@ namespace {
 bool attack_type::get_special_bool(const std::string& special, bool simple_check) const
 {
 	{
-		std::vector<const config*> list;
-		if ( get_special_children(list, specials_, special, simple_check) ) {
+		std::vector<const config*> tag_list, id_list;
+		if ( get_special_children(tag_list, id_list, specials_, special, simple_check) ) {
 			return true;
 		}
 		// If we make it to here, then either list.empty() or !simple_check.
 		// So if the list is not empty, then this is not a simple check and
 		// we need to check each special in the list to see if any are active.
-		for(const config* entry : list) {
-			if ( special_active(*entry, AFFECT_SELF) ) {
+		for(const config* entry : tag_list) {
+			if ( special_active(*entry, AFFECT_SELF, special) ) {
 				return true;
 			}
 		}
 	}
+
 	// Skip checking the opponent's attack?
 	if ( simple_check || !other_attack_ ) {
 		return false;
 	}
 
-	std::vector<const config*> list;
-	get_special_children(list, other_attack_->specials_, special);
-	for(const config* entry : list) {
-		if ( other_attack_->special_active(*entry, AFFECT_OTHER) ) {
+	std::vector<const config*> tag_list, id_list;
+	get_special_children(tag_list, id_list, other_attack_->specials_, special);
+	for(const config* entry : tag_list) {
+		if ( other_attack_->special_active(*entry, AFFECT_OTHER, special) ) {
```

Comment:
Thanks for rebasing! It seems the `id_list` return value is never used, now? Though if we _did_ use it, we'd need get_special_children to return the value of sp.key along with it, so we could pass it to special_active.

Reply:
Yes, I noticed that we are not using `id_list` in this case.

> Though if we _did_ use it, we'd need get_special_children to return the value of sp.key along with it, so we could pass it to special_active.

I'm not sure I understand what you mean. Since tag_list contains the specials where the tag name matches, wouldn't that mean that `special == sp.key` anyway?

Reply:
When iterating over id_list, the third parameter to special_active should be the tag name, like it is in all other cases, not the id.

Reply:
You're talking about within the `get_special_bool` function, right? It's iterating through tag_list both times, not id_list. I thought that the point was that get_special_bool should be checking the tag name instead of the id.

Reply:
No, it's special_active that should check the tag name rather than the id. (The PR already makes it do that.) `get_special_bool` should keep doing what it does in 1.14/master, but pass a tag name, not an id, for special_active's third parameter.

As @gfgtdf said, we can later split get_special_bool into two functions, one for "tag matches" and one for "id matches", but I don't think that has to be part of this PR.

Reply:
Thanks, I think I get it now. So within `get_special_bool`, I should pass the tag_name instead of the id (represented by `special`) to special_active, and this tag_name is to be retrieved from the call to `get_special_children`. 

I had already made changes to get_special_children to return two lists, one for matching tags and the other for matching ids. In this case, I should revert back to what it was originally, one list for matching tags or ids, and also change it to return the matching tag_name, right?

Thanks for your patience!

Reply:
The important thing is to return the matching tag name for tags that were matched because of equal id, because get_special_bool needs to pass them to special_active. Personally, I would leave get_special_bool returning two separate lists because that will make the future changes to split get_special_bool into two functions (one for matching by id and one for matching by tag name) smaller.

Reply:
I've modified `get_special_children` to accept one more parameter, a string `tag_name`, which will store the matching tag name when a matching special is found (either sp.key or sp.cfg["id"] == id)
Then, in `get_special_bool`, now we have the tag_name from the call to `get_special_children`, and we pass that to `special_active` instead of `special`

Reply:
I assume that would work in practice since usually there would be either one id-match or one or more tagname-matches, but I don't want to introduce technical debt. I think it would be better to explicitly couple config objects with their tag names, rather than return two vectors and one string that's related to one of the elements in one of the vectors. Maybe return a vector of `struct { std::string tag_name; config config; }` objects instead of a vector of `config*`? Anyone else has a suggestion?

Reply:
While you definitely have a point, I think that's a much larger project that's out of scope for this PR.

Reply:
@jostephd  I do agree that the current approach is a bit messy. If we were to do that, then we would define the struct in abilities.hpp right?

@CelticMinstrel in this case, shall we keep what we have for this current PR?

Reply:
> @jostephd I do agree that the current approach is a bit messy. If we were to do that, then we would define the struct in abilities.hpp right?

You should declare/define the struct in the narrowest scope possible: private to one cpp file in possible, or declared in the hpp file (`struct foo;`) and defined in the cpp file (`struct foo { ... }`).

I'm unassigning myself since I might not have be able to iterate/review timely. I'll try to keep tabs on this PR but other developers please feel free to take over.